keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md

```diff
@@ -99,105 +101,167 @@ A developer has a Windows web app that they would like to deploy in a container
+  - Populate the contents of the credspec JSON annotations or securityContext fields if not already specified.
+  - Check in a deep-equal fashion, the contents of the credspec JSON annotations or securityContext fields if already specified. If the check fails, pod creation will be failed with 400: BadRequest
 
-During pod updates, changes to the `pod.[alpha|beta].kubernetes.io/windows-gmsa-config-map` annotation will be blocked by `ValidatePodSpecificAnnotationUpdates`. `ValidatePodUpdate` already ensures a pod's `securityContext` cannot be updated. So no further checks/blocks are necessary to prevent update operations around the reference to the credspec configmap on pods in GA.
+Next, GMSAAuthorizer will ensure the service account specified for the pod is authorized for a special `use` verb on the GMSACredSpec objects whose name is specified through the annotations or securityContext fields. GMSAAuthorizer will generate custom `AttributesRecord`s with `verb` set to `use`, `name` set to the GMSACredSpec object and `user` set to the service account of the pod. Finally, the `AttributesRecord`s will be passed to authorizers to check against RBAC configurations. A failure from the authorizes results in a 403: Forbidden.
 
-#### Changes in CRI's WindowsContainerSecurityContext struct
-A new field `CredentialSpec String` will be added to `WindowsContainerSecurityContext`. It will be populated by `generateWindowsContainerConfig` for Windows containers with the contents of key `gmsa-cred-spec` in the GMSA configmap reference from the pod (specified either through the annotation `pod.[alpha|beta].kubernetes.io/windows-gmsa-config-map` during alpha/beta stages or `securityContext.windows.credentialSpecConfig` in the pod spec at GA).
+During pod updates, changes to the credspec annotations will be blocked by GMSAAuthorizer and failed with 400: BadRequest.
+
+If the webhook is not installed and configured, no authorization checks will be performed on the contents of the credspec JSON. This will allow arbitrary credspec JSON to be specified for pods/containers and sent down to the container runtime.
+
+#### Changes in CRI's WindowsContainerSecurityContext struct and DetermineEffectiveSecurityContext
+
+A new field `CredentialSpec String` will be added to `WindowsContainerSecurityContext`. 
+
+`DetermineEffectiveSecurityContext` will be enhanced to populate `CredentialSpec` for Windows containers based on credspec specification for containers and pod. It will need to obtain the credspec contents from the `pod.alpha.kubernetes.io/windows-gmsa-credspec` and `container.windows-gmsa-credspec.alpha.kubernetes.io/container-name` annotations during the Alpha phase. The enhancements will be enabled under a feature flag during Alpha phase of the feature.
 
 #### Changes in Dockershim
```

Comment:
how would this change with the containerD changes down the pipeline

Reply:
Yes, please add a brief section about how this will be implemented in containerd.

Reply:
Done